File: src/providers/frp/domain/proxy.ts
import { i18n } from '@/i18n'
import { hydratePlugin, type PluginOptions } from './plugin'

export type ProxyType = 'tcp' | 'udp' | 'http' | 'https' | 'tcpmux' | 'stcp' | 'sudp' | 'xtcp'

export interface ProxyConfig {
    type: ProxyType
    name: string
    localIP: string
    localPort: number | null
    remotePort: number | null
    customDomains: string[]
    subdomain: string
    secretKey: string
    multiplexer: string
    httpUser: string
    httpPassword: string
    hostHeaderRewrite: string
    requestHeaders: Record<string, string>
    responseHeaders: Record<string, string>
    locations: string[]
    routeByHTTPUser: string
    allowUsers: string[]
    loadBalancer: { group: string; groupKey: string }
    healthCheck: {
        type: string
        timeoutSeconds: number
        maxFailed: number
        intervalSeconds: number
        path: string
        httpHeaders: { name: string; value: string }[]
    }
    transport: {
        useEncryption: boolean
        useCompression: boolean
        bandwidthLimit: string
        bandwidthLimitMode: string
        proxyProtocolVersion: string
    }
    plugin?: PluginOptions | null
    metadatas: Record<string, string> | null
    annotations: Record<string, string> | null
}

export interface ProfileProxy {
    id: string
    enabled: boolean
    config: ProxyConfig
}

/**
 * 中性默认值：所有「开关型」字段都取关闭/缺省（healthCheck.type=''、transport.bandwidthLimitMode=''、
 * 端口 null）。后端对这些空值做 skip_serializing_if 省略，因此 hydrate 回填它们不会改写
 * 已存储的配置，也不会触发虚假的「待重启」。新建隧道的按类型预填在 newProxyConfig 里做。
 */
export function defaultProxyConfig(type: ProxyType = 'tcp'): ProxyConfig {
    return {
        type,
        name: '',
        localIP: '127.0.0.1',
        localPort: null,
        remotePort: null,
        customDomains: [],
        subdomain: '',
        secretKey: '',
        multiplexer: '',
        httpUser: '',
        httpPassword: '',
        hostHeaderRewrite: '',
        requestHeaders: {},
        responseHeaders: {},
        locations: [],
        routeByHTTPUser: '',
        allowUsers: [],
        loadBalancer: { group: '', groupKey: '' },
        healthCheck: {
            // ''=不启用；秒数是启用后的预填值，type 为空时后端整体省略
            type: '',
            timeoutSeconds: 3,
            maxFailed: 3,
            intervalSeconds: 10,
            path: '',
            httpHeaders: [],
        },
        transport: {
            useEncryption: false,
            useCompression: false,
            bandwidthLimit: '',
            bandwidthLimitMode: '',
            proxyProtocolVersion: '',
        },
        metadatas: null,
        annotations: null,
    }
}

/** 各隧道类型的新建预填（只作用于新建/切换类型，不参与已存配置的 hydrate）。 */
const TYPE_PRESETS: Partial<Record<ProxyType, Partial<ProxyConfig>>> = {
    http: { localPort: 80 },
    https: { localPort: 443 },
    // frp 的 tcpmux 目前只支持 httpconnect，且编辑器没有该字段的输入框，必须预填。
    tcpmux: { multiplexer: 'httpconnect' },
}

export function newProxyConfig(type: ProxyType = 'tcp'): ProxyConfig {
    return { ...defaultProxyConfig(type), ...TYPE_PRESETS[type] }
}

export function newProfileProxy(type: ProxyType = 'tcp'): ProfileProxy {
    return {
        id: crypto.randomUUID(),
        enabled: true,
        config: newProxyConfig(type),
    }
}

export function hydrateProxyConfig(raw: any): ProxyConfig {
    const base = defaultProxyConfig(raw?.type)
    if (!raw) return base
    return {
        ...base,
        ...raw,
        // 0 不是合法的本地/远端端口（曾有版本把默认 0 写进存档）：归一成 null，
        // 序列化时按「未设置」省略，恢复 frp 期望的形态。
        localPort: raw.localPort || null,
        remotePort: raw.remotePort || null,
        customDomains: raw.customDomains ?? [],
        requestHeaders: raw.requestHeaders ?? {},
        responseHeaders: raw.responseHeaders ?? {},
        locations: raw.locations ?? [],
        allowUsers: raw.allowUsers ?? [],
        loadBalancer: { ...base.loadBalancer, ...raw.loadBalancer },
        healthCheck: { ...base.healthCheck, ...raw.healthCheck },
        transport: { ...base.transport, ...raw.transport },
        // 后端会省略插件里的空字段（skip_serializing_if），编辑器按「键是否存在」决定
        // 渲染哪些输入框，必须补全该插件类型的全部键，否则字段静默消失。
        plugin: raw.plugin ? hydratePlugin(raw.plugin) : raw.plugin,
        metadatas: raw.metadatas ?? null,
        annotations: raw.annotations ?? null,
    }
}

export function hydrateProfileProxy(raw: any): ProfileProxy {
    return {
        id: raw?.id ?? crypto.randomUUID(),
        enabled: raw?.enabled ?? true,
        config: hydrateProxyConfig(raw?.config),
    }
}

/**
 * 列表里展示的「远端」描述的数据结构。
 * Domain 层只返回结构化数据，UI 层负责国际化翻译。
 */
export type RemoteSummaryData =
    | { kind: 'port'; port: number }
    | { kind: 'dynamicPort' }
    | { kind: 'domains'; domains: string[] }
    | { kind: 'secret'; set: boolean }

/** 返回远端配置的结构化数据，供 UI 层格式化 */
export function remoteSummaryData(c: ProxyConfig): RemoteSummaryData {
    switch (c.type) {
        case 'tcp':
        case 'udp':
            return c.remotePort ? { kind: 'port', port: c.remotePort } : { kind: 'dynamicPort' }
        case 'http':
        case 'https':
        case 'tcpmux':
            const domains = [...c.customDomains, ...(c.subdomain ? [c.subdomain] : [])]
            return { kind: 'domains', domains }
        case 'stcp':
        case 'sudp':
        case 'xtcp':
            return { kind: 'secret', set: !!c.secretKey }
        default:
            return { kind: 'dynamicPort' }
    }
}

/** 列表里展示的「远端」描述（带 i18n）。仅在 UI 渲染时调用，保证 i18n 已初始化。 */
export function remoteSummary(c: ProxyConfig): string {
    const data = remoteSummaryData(c)
    const t = i18n.global.t

    switch (data.kind) {
        case 'port':
            return `:${data.port}`
        case 'dynamicPort':
            return t('proxySummary.dynamicPort')
        case 'domains':
            return data.domains.join(', ') || '—'
        case 'secret':
            return data.set ? t('proxySummary.secretSet') : t('proxySummary.secretUnset')
    }
}

/** 列表里展示的「本地」描述。 */
export function localSummary(c: ProxyConfig): string {
    if (c.plugin) return `plugin: ${c.plugin.type}`
    const ip = c.localIP || '127.0.0.1'
    return c.localPort ? `${ip}:${c.localPort}` : ip
}

File: src-tauri/src/providers/frp/domain/proxy.rs
//! 隧道（代理）配置，对应 frp v1 的 ProxyConfigurer 各实现。字段名与 frp json tag 对齐。
use super::types::{gen_id, is_false};
use super::PluginOptions;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

type Map = HashMap<String, String>;

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HeaderOps {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub set: Option<Map>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HttpHeader {
    pub name: String,
    pub value: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProxyTransport {
    #[serde(default, skip_serializing_if = "is_false")]
    pub use_encryption: bool,
    #[serde(default, skip_serializing_if = "is_false")]
    pub use_compression: bool,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub bandwidth_limit: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub bandwidth_limit_mode: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub proxy_protocol_version: String,
}

impl ProxyTransport {
    fn is_empty(&self) -> bool {
        !self.use_encryption
            && !self.use_compression
            && self.bandwidth_limit.is_empty()
            && self.bandwidth_limit_mode.is_empty()
            && self.proxy_protocol_version.is_empty()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HealthCheck {
    /// "tcp" | "http" | ""（空=不启用）
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub r#type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_seconds: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_failed: Option<i32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub interval_seconds: Option<i32>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub path: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub http_headers: Vec<HttpHeader>,
}

impl HealthCheck {
    fn is_empty(&self) -> bool {
        self.r#type.is_empty()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct LoadBalancer {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub group: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub group_key: String,
}

impl LoadBalancer {
    fn is_empty(&self) -> bool {
        self.group.is_empty() && self.group_key.is_empty()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct NatTraversal {
    #[serde(default, skip_serializing_if = "is_false")]
    pub disable_assisted_addrs: bool,
}

/// 所有隧道共有字段（frp ProxyBaseConfig + ProxyBackend，去掉 type/enabled）。
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProxyBase {
    pub name: String,
    #[serde(default, skip_serializing_if = "ProxyTransport::is_empty")]
    pub transport: ProxyTransport,
    #[serde(default, skip_serializing_if = "HealthCheck::is_empty")]
    pub health_check: HealthCheck,
    #[serde(default, skip_serializing_if = "LoadBalancer::is_empty")]
    pub load_balancer: LoadBalancer,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metadatas: Option<Map>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub annotations: Option<Map>,
    #[serde(rename = "localIP", default, skip_serializing_if = "String::is_empty")]
    pub local_ip: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub local_port: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub plugin: Option<PluginOptions>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TcpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remote_port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UdpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub remote_port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HttpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub custom_domains: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subdomain: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub locations: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub http_user: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub http_password: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub host_header_rewrite: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_headers: Option<HeaderOps>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub response_headers: Option<HeaderOps>,
    #[serde(
        rename = "routeByHTTPUser",
        default,
        skip_serializing_if = "String::is_empty"
    )]
    pub route_by_http_user: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct HttpsProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub custom_domains: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subdomain: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TcpmuxProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub custom_domains: Vec<String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub subdomain: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub http_user: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub http_password: String,
    #[serde(
        rename = "routeByHTTPUser",
        default,
        skip_serializing_if = "String::is_empty"
    )]
    pub route_by_http_user: String,
    /// 目前仅 "httpconnect"
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub multiplexer: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct StcpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub secret_key: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allow_users: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SudpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub secret_key: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allow_users: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct XtcpProxy {
    #[serde(flatten)]
    pub base: ProxyBase,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub secret_key: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allow_users: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub nat_traversal: Option<NatTraversal>,
}

/// 隧道配置：内部按 "type" 标签区分，序列化即 frp 期望的扁平对象。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ProxyConfig {
    Tcp(TcpProxy),
    Udp(UdpProxy),
    Http(HttpProxy),
    Https(HttpsProxy),
    Tcpmux(TcpmuxProxy),
    Stcp(StcpProxy),
    Sudp(SudpProxy),
    Xtcp(XtcpProxy),
}

impl ProxyConfig {
    pub fn base(&self) -> &ProxyBase {
        match self {
            ProxyConfig::Tcp(p) => &p.base,
            ProxyConfig::Udp(p) => &p.base,
            ProxyConfig::Http(p) => &p.base,
            ProxyConfig::Https(p) => &p.base,
            ProxyConfig::Tcpmux(p) => &p.base,
            ProxyConfig::Stcp(p) => &p.base,
            ProxyConfig::Sudp(p) => &p.base,
            ProxyConfig::Xtcp(p) => &p.base,
        }
    }

    pub fn base_mut(&mut self) -> &mut ProxyBase {
        match self {
            ProxyConfig::Tcp(p) => &mut p.base,
            ProxyConfig::Udp(p) => &mut p.base,
            ProxyConfig::Http(p) => &mut p.base,
            ProxyConfig::Https(p) => &mut p.base,
            ProxyConfig::Tcpmux(p) => &mut p.base,
            ProxyConfig::Stcp(p) => &mut p.base,
            ProxyConfig::Sudp(p) => &mut p.base,
            ProxyConfig::Xtcp(p) => &mut p.base,
        }
    }
}

/// 持久化层的隧道：在 frp 配置外面包一层 id/enabled，供 UI 使用；生成运行配置时只写 config。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfileProxy {
    #[serde(default = "gen_id")]
    pub id: String,
    #[serde(default = "default_true")]
    pub enabled: bool,
    pub config: ProxyConfig,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn http_proxy_roundtrip() {
        let base = ProxyBase {
            name: "web".into(),
            local_ip: "127.0.0.1".into(),
            local_port: Some(80),
            ..Default::default()
        };
        let p = ProxyConfig::Http(HttpProxy {
            base,
            subdomain: "app".into(),
            custom_domains: vec!["a.com".into()],
            ..Default::default()
        });
        let j = serde_json::to_string(&p).unwrap();
        assert!(j.contains("\"type\":\"http\""), "json={j}");
        assert!(j.contains("\"name\":\"web\""));
        assert!(j.contains("\"localIP\":\"127.0.0.1\""));
        assert!(j.contains("\"subdomain\":\"app\""));
        // 不该出现 tcp 专属字段
        assert!(!j.contains("remotePort"));

        let back: ProxyConfig = serde_json::from_str(&j).unwrap();
        match back {
            ProxyConfig::Http(h) => {
                assert_eq!(h.base.name, "web");
                assert_eq!(h.base.local_port, Some(80));
                assert_eq!(h.subdomain, "app");
            }
            _ => panic!("Expected Http variant, got {:?}", back),
        }
    }

    #[test]
    fn profile_proxy_wrapper_roundtrip() {
        let p = ProfileProxy {
            id: "x".into(),
            enabled: true,
            config: ProxyConfig::Tcp(TcpProxy {
                base: ProxyBase {
                    name: "ssh".into(),
                    local_port: Some(22),
                    ..Default::default()
                },
                remote_port: Some(6000),
            }),
        };
        let j = serde_json::to_string(&p).unwrap();
        let back: ProfileProxy = serde_json::from_str(&j).unwrap();
        assert_eq!(back.id, "x");
        assert!(matches!(back.config, ProxyConfig::Tcp(_)));
    }
}
